app/jobs/store_version_contents_job.rb

```diff
@@ -0,0 +1,39 @@
+class StoreVersionContentsJob < ApplicationJob
+  include GoodJob::ActiveJobExtensions::Concurrency
+  good_job_control_concurrency_with total_limit: 1, key: -> { "store-contents-#{version_arg.full_name}" }
+  queue_as :default
+  self.queue_adapter = :good_job
+
+  class VersionNotIndexed < RuntimeError; end
+  class GemNotFound < RuntimeError; end
+
+  discard_on ActiveJob::DeserializationError
+  discard_on Gem::Package::FormatError, Gem::Security::Exception
+
+  retry_on VersionNotIndexed, wait: :exponentially_longer, attempts: 5
+  retry_on GemNotFound, wait: :exponentially_longer, attempts: 5
+
+  rescue_from(GemNotFound, Gem::Package::FormatError, Gem::Security::Exception) do |error|
+    version = version_arg.full_name
+    logger.error "Storing gem contents for #{version} failed", error
+    Rails.error.report error, context: { version: version }, handled: false
+    raise error
+  end
+
+  def version_arg
+    arguments.first[:version]
+  end
+
+  def perform(version:)
+    raise VersionNotIndexed, "Version #{version&.full_name.inspect} is not indexed" unless version&.indexed?
+    logger.info "Storing gem contents for #{version.full_name}"
+
+    gem = RubygemFs.instance.get("gems/#{version.full_name}.gem")
```

Comment:
Does this load whole gem into memory? There are even 450MB gems - https://rubygems.org/gems/emojidex-rasters/versions/1.0.34.

Reply:
It will. As far as I know, we regularly load the entire gem into memory on push, so it's not an unreasonable lift. The way it's processed the max memory usage should be gem + largest file + overhead of stream processing.

Reply:
OK, I was thinking about streaming file to temp file first and reading it from disk. I'll take a look later on this. We can do that optionally only for huge files at some point.